packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
+                    visited: false,
+                    referencedRoutes: new Set<string>(),
+                };
+            }
+
+            const visited: boolean = references.visited;
+            // Keeps track of whether the entry for this data store in the dataStoreReferencesMap changes.
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
+                assert(references.referencedRoutes);
+                references.referencedRoutes.add(dataStoreIdRelativePath);
+                dataStoreEntryChanged = true;
+            }
+
+            // If this data store has not been visited yet, mark it as visited.
+            if (!references.visited) {
+                references.visited = true;
+                dataStoreEntryChanged = true;
+            }
+
+            // If the entry has changed, update it in the dataStoreReferencesMap.
+            if (dataStoreEntryChanged) {
+                dataStoreReferencesMap.set(dataStoreId, references);
+            }
+
+            // If this data store has been already visited, ignore it.
+            if (visited) {
+                continue;
+            }
+
+            // Get the routes referenced by this data store and add them to the referencedRoutesStack.
+            const referencedRoutes = this.getDataStoreReferencedRoutes(dataStoreId);
+            referencedRoutesStack.push(...referencedRoutes);
+        }
+
+        const deletedDataStores: string[] = [];
+        // Now we have a map of all the data stores that are referenced and the routes that are referenced in it.
+        // Iterate through all the data stores in this runtime and:
+        // 1. If they are not referneced, mark them for collection.
+        // 2. If they are referenced, pass the list of routes in them that are referenced so it can collect
+        //    unreferenced objects.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            // Ignore the scheduler.
+            if (value.id === taskSchedulerId) {
+                return;
+            }
+
+            // If the data store does exists in the dataStoreReferencesMap, it is no referenced. Mark it for collection.
+            // Else, get the list of the referenced routes in it and ask it to collect the unreferenced objects.
+            if (!dataStoreReferencesMap.has(value.id)) {
+                deletedDataStores.push(value.id);
+                value.setReferenced(false);
+            } else {
+                const references = dataStoreReferencesMap.get(value.id);
+                assert(references);
+                value.setReferenced(true);
+                value.collectUnreferencedFluidObjects([...references.referencedRoutes]);
```

Comment:
I'd change name to notifyUsedRoutes() or something like that, otherwise name and argument are at odds with each other